In [1]:
import pandas as pd
from pandas import PeriodIndex

In [2]:
df_vehicle_crash = pd.read_csv(r"C:\Users\franc\OneDrive\Documentos\Henry\proyectos\PROYECTO_GRUPAL\datasets_new\motor-vehicle-collisions-crashes.csv")

C:\Users\franc\AppData\Local\Temp\ipykernel_23724\1714604262.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  df_vehicle_crash = pd.read_csv(r"C:\Users\franc\OneDrive\Documentos\Henry\proyectos\PROYECTO_GRUPAL\datasets_new\motor-vehicle-collisions-crashes.csv")


In [3]:
df_vehicle_crash.sort_values(["CRASH_DATE", "CRASH_TIME"], inplace = True)

df_vehicle_crash.reset_index(drop = True, inplace = True)

df_vehicle_crash.head()

CRASH_DATE CRASH_TIME    BOROUGH ZIP_CODE   LATITUDE  LONGITUDE  \
0  01/01/2013       0:01     QUEENS    11369  40.761609 -73.882678   
1  01/01/2013       0:05  MANHATTAN    10018  40.757732 -73.996852   
2  01/01/2013       0:07  MANHATTAN    10002  40.720968 -73.982168   
3  01/01/2013       0:20  MANHATTAN    10075  40.772074 -73.949939   
4  01/01/2013       0:25     QUEENS    11419  40.683430 -73.826601   

                    LOCATION                    ON_STREET_NAME  \
0  (40.7616085, -73.8826782)  86 STREET                          
1  (40.7577322, -73.9968524)  10 AVENUE                          
2  (40.7209677, -73.9821677)  EAST HOUSTON STREET                
3  (40.7720736, -73.9499388)  EAST 80 STREET                     
4    (40.68343, -73.8266015)  115 STREET                         

                  CROSS_STREET_NAME OFF_STREET_NAME  ...  \
0  30 AVENUE                                    NaN  ...   
1  WEST 39 STREET                               NaN  ...   
2  RIDGE STREET                                 NaN  ...   
3  YORK AVENUE                                  NaN  ...   
4  107 AVENUE                                   NaN  ...   

   CONTRIBUTING_FACTOR_VEHICLE_2  CONTRIBUTING_FACTOR_VEHICLE_3  \
0                    Unspecified                            NaN   
1                    Unspecified                    Unspecified   
2    Traffic Control Disregarded                            NaN   
3                    Unspecified                            NaN   
4                    Unspecified                    Unspecified   

   CONTRIBUTING_FACTOR_VEHICLE_4  CONTRIBUTING_FACTOR_VEHICLE_5  COLLISION_ID  \
0                            NaN                            NaN        281035   
1                            NaN                            NaN         18025   
2                            NaN                            NaN         12712   
3                            NaN                            NaN         47038   
4                    Unspecified                            NaN        229479   

   VEHICLE_TYPE_CODE_1            VEHICLE_TYPE_CODE_2  \
0    PASSENGER VEHICLE  SPORT UTILITY / STATION WAGON   
1    PASSENGER VEHICLE                        UNKNOWN   
2                 TAXI              PASSENGER VEHICLE   
3                 TAXI                           TAXI   
4              UNKNOWN              PASSENGER VEHICLE   

             VEHICLE_TYPE_CODE_3 VEHICLE_TYPE_CODE_4 VEHICLE_TYPE_CODE_5  
0                            NaN                 NaN                 NaN  
1                        UNKNOWN                 NaN                 NaN  
2                            NaN                 NaN                 NaN  
3                            NaN                 NaN                 NaN  
4  SPORT UTILITY / STATION WAGON   PASSENGER VEHICLE                 NaN  

[5 rows x 29 columns]

In [4]:
df_vehicle_crash.columns

Index(['CRASH_DATE', 'CRASH_TIME', 'BOROUGH', 'ZIP_CODE', 'LATITUDE',
       'LONGITUDE', 'LOCATION', 'ON_STREET_NAME', 'CROSS_STREET_NAME',
       'OFF_STREET_NAME', 'NUMBER_OF_PERSONS_INJURED',
       'NUMBER_OF_PERSONS_KILLED', 'NUMBER_OF_PEDESTRIANS_INJURED',
       'NUMBER_OF_PEDESTRIANS_KILLED', 'NUMBER_OF_CYCLIST_INJURED',
       'NUMBER_OF_CYCLIST_KILLED', 'NUMBER_OF_MOTORIST_INJURED',
       'NUMBER_OF_MOTORIST_KILLED', 'CONTRIBUTING_FACTOR_VEHICLE_1',
       'CONTRIBUTING_FACTOR_VEHICLE_2', 'CONTRIBUTING_FACTOR_VEHICLE_3',
       'CONTRIBUTING_FACTOR_VEHICLE_4', 'CONTRIBUTING_FACTOR_VEHICLE_5',
       'COLLISION_ID', 'VEHICLE_TYPE_CODE_1', 'VEHICLE_TYPE_CODE_2',
       'VEHICLE_TYPE_CODE_3', 'VEHICLE_TYPE_CODE_4', 'VEHICLE_TYPE_CODE_5'],
      dtype='object')

In [5]:
# me quedo con las columnas de interes
df_vehicle_crash_clean = df_vehicle_crash[[ 'COLLISION_ID', 'CRASH_DATE', 'CRASH_TIME', 'BOROUGH', 'ZIP_CODE', 'LATITUDE', 
                                            'LONGITUDE', 'LOCATION', 'ON_STREET_NAME', 'CROSS_STREET_NAME', 
                                            'OFF_STREET_NAME', 'NUMBER_OF_PERSONS_INJURED', 'NUMBER_OF_PERSONS_KILLED', 
                                            'NUMBER_OF_PEDESTRIANS_INJURED', 'NUMBER_OF_PEDESTRIANS_KILLED', 
                                            'NUMBER_OF_CYCLIST_INJURED', 'NUMBER_OF_CYCLIST_KILLED', 'NUMBER_OF_MOTORIST_INJURED',
                                            'NUMBER_OF_MOTORIST_KILLED', 'CONTRIBUTING_FACTOR_VEHICLE_1',
                                            'CONTRIBUTING_FACTOR_VEHICLE_2', 'CONTRIBUTING_FACTOR_VEHICLE_3',
                                            'CONTRIBUTING_FACTOR_VEHICLE_4', 'CONTRIBUTING_FACTOR_VEHICLE_5']]

In [6]:
# hago la captura del dia de la semana
dates = df_vehicle_crash_clean["CRASH_DATE"]
idx = pd.PeriodIndex(dates, freq = "D") #configuro la frecuencia para que sea por dia
week_day = idx.dayofweek #asigno la configuracion de numero de dia
# week_day
df_vehicle_crash_clean["WEEK_DAY"] = week_day

# lunes = 0
# martes = 1
# miercoles = 2
# jueves = 3
# viernes = 4

C:\Users\franc\AppData\Local\Temp\ipykernel_23724\1537956710.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vehicle_crash_clean["WEEK_DAY"] = week_day


In [10]:
# hago la captura mes
months = df_vehicle_crash_clean["CRASH_DATE"]
idx = pd.PeriodIndex(dates, freq = "M") #configuro la frecuencia para que sea por dia
month_num = idx.month #asigno la configuracion de numero de dia
# week_day
df_vehicle_crash_clean["MONTH"] = month_num

In [12]:
# hago la captura año
years = df_vehicle_crash_clean["CRASH_DATE"]
idx = pd.PeriodIndex(dates, freq = "Y") #configuro la frecuencia para que sea por dia
year = idx.year #asigno la configuracion de numero de dia
# week_day
df_vehicle_crash_clean["YEAR"] = year

C:\Users\franc\AppData\Local\Temp\ipykernel_23724\3738290553.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_vehicle_crash_clean["YEAR"] = year


In [13]:
# pongo el dia de la semana entre las primeras columnas
df_vehicle_crash_clean = df_vehicle_crash_clean[[   'COLLISION_ID', "CRASH_DATE", "CRASH_TIME", "WEEK_DAY", "MONTH", "YEAR", "BOROUGH", "LATITUDE", 
                                                    "LONGITUDE", "ON_STREET_NAME", "CROSS_STREET_NAME", "OFF_STREET_NAME",
                                                    "NUMBER_OF_PERSONS_INJURED", "NUMBER_OF_PERSONS_KILLED", 
                                                    "NUMBER_OF_PEDESTRIANS_INJURED", "NUMBER_OF_PEDESTRIANS_KILLED", 
                                                    "NUMBER_OF_CYCLIST_INJURED", "NUMBER_OF_CYCLIST_KILLED", 
                                                    "NUMBER_OF_MOTORIST_INJURED", "NUMBER_OF_MOTORIST_KILLED", 
                                                    'CONTRIBUTING_FACTOR_VEHICLE_1', 'CONTRIBUTING_FACTOR_VEHICLE_2', 
                                                    'CONTRIBUTING_FACTOR_VEHICLE_3', 'CONTRIBUTING_FACTOR_VEHICLE_4', 
                                                    'CONTRIBUTING_FACTOR_VEHICLE_5'  ]]

df_vehicle_crash_clean.head()

COLLISION_ID  CRASH_DATE CRASH_TIME  WEEK_DAY  MONTH  YEAR    BOROUGH  \
0        281035  01/01/2013       0:01         1      1  2013     QUEENS   
1         18025  01/01/2013       0:05         1      1  2013  MANHATTAN   
2         12712  01/01/2013       0:07         1      1  2013  MANHATTAN   
3         47038  01/01/2013       0:20         1      1  2013  MANHATTAN   
4        229479  01/01/2013       0:25         1      1  2013     QUEENS   

    LATITUDE  LONGITUDE                    ON_STREET_NAME  ...  \
0  40.761609 -73.882678  86 STREET                         ...   
1  40.757732 -73.996852  10 AVENUE                         ...   
2  40.720968 -73.982168  EAST HOUSTON STREET               ...   
3  40.772074 -73.949939  EAST 80 STREET                    ...   
4  40.683430 -73.826601  115 STREET                        ...   

  NUMBER_OF_PEDESTRIANS_KILLED NUMBER_OF_CYCLIST_INJURED  \
0                            0                         0   
1                            0                         0   
2                            0                         0   
3                            0                         0   
4                            0                         0   

   NUMBER_OF_CYCLIST_KILLED  NUMBER_OF_MOTORIST_INJURED  \
0                         0                           0   
1                         0                           0   
2                         0                           1   
3                         0                           0   
4                         0                           0   

   NUMBER_OF_MOTORIST_KILLED  CONTRIBUTING_FACTOR_VEHICLE_1  \
0                          0                    Unspecified   
1                          0               Brakes Defective   
2                          0            Driver Inexperience   
3                          0                    Unspecified   
4                          0                    Unspecified   

   CONTRIBUTING_FACTOR_VEHICLE_2  CONTRIBUTING_FACTOR_VEHICLE_3  \
0                    Unspecified                            NaN   
1                    Unspecified                    Unspecified   
2    Traffic Control Disregarded                            NaN   
3                    Unspecified                            NaN   
4                    Unspecified                    Unspecified   

   CONTRIBUTING_FACTOR_VEHICLE_4  CONTRIBUTING_FACTOR_VEHICLE_5  
0                            NaN                            NaN  
1                            NaN                            NaN  
2                            NaN                            NaN  
3                            NaN                            NaN  
4                    Unspecified                            NaN  

[5 rows x 25 columns]

In [14]:
# arreglo errores de tipeo
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_1"].replace("Illnes", "Illness", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_1"].replace("Cell Phone (hand-Held)", "Cell Phone (hand-held)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_1"].replace("Drugs (Illegal)", "Drugs (illegal)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_1"].replace("Reaction to Other Uninvolved Vehicle", "Reaction to Uninvolved Vehicle", inplace = True)

df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_2"].replace("Illnes", "Illness", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_2"].replace("Cell Phone (hand-Held)", "Cell Phone (hand-held)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_2"].replace("Drugs (Illegal)", "Drugs (illegal)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_2"].replace("Reaction to Other Uninvolved Vehicle", "Reaction to Uninvolved Vehicle", inplace = True)

df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_3"].replace("Illnes", "Illness", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_3"].replace("Cell Phone (hand-Held)", "Cell Phone (hand-held)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_3"].replace("Drugs (Illegal)", "Drugs (illegal)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_3"].replace("Reaction to Other Uninvolved Vehicle", "Reaction to Uninvolved Vehicle", inplace = True)

df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_4"].replace("Illnes", "Illness", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_4"].replace("Cell Phone (hand-Held)", "Cell Phone (hand-held)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_4"].replace("Drugs (Illegal)", "Drugs (illegal)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_4"].replace("Reaction to Other Uninvolved Vehicle", "Reaction to Uninvolved Vehicle", inplace = True)

df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_5"].replace("Illnes", "Illness", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_5"].replace("Cell Phone (hand-Held)", "Cell Phone (hand-held)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_5"].replace("Drugs (Illegal)", "Drugs (illegal)", inplace = True)
df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_5"].replace("Reaction to Other Uninvolved Vehicle", "Reaction to Uninvolved Vehicle", inplace = True)

In [15]:
df_vehicle_crash_clean

COLLISION_ID  CRASH_DATE CRASH_TIME  WEEK_DAY  MONTH  YEAR  \
0              281035  01/01/2013       0:01         1      1  2013   
1               18025  01/01/2013       0:05         1      1  2013   
2               12712  01/01/2013       0:07         1      1  2013   
3               47038  01/01/2013       0:20         1      1  2013   
4              229479  01/01/2013       0:25         1      1  2013   
...               ...         ...        ...       ...    ...   ...   
1918532       4492926  12/31/2021       9:00         4     12  2021   
1918533       4491580  12/31/2021       9:16         4     12  2021   
1918534       4491634  12/31/2021       9:17         4     12  2021   
1918535       4493202  12/31/2021       9:29         4     12  2021   
1918536       4491080  12/31/2021       9:50         4     12  2021   

           BOROUGH   LATITUDE  LONGITUDE                    ON_STREET_NAME  \
0           QUEENS  40.761609 -73.882678  86 STREET                          
1        MANHATTAN  40.757732 -73.996852  10 AVENUE                          
2        MANHATTAN  40.720968 -73.982168  EAST HOUSTON STREET                
3        MANHATTAN  40.772074 -73.949939  EAST 80 STREET                     
4           QUEENS  40.683430 -73.826601  115 STREET                         
...            ...        ...        ...                               ...   
1918532     QUEENS  40.769684 -73.821650                         32 AVENUE   
1918533        NaN  40.839970 -73.877920               EAST TREMONT AVENUE   
1918534     QUEENS        NaN        NaN                        138 Street   
1918535        NaN  40.882630 -73.839516                 BAYCHESTER AVENUE   
1918536      BRONX  40.816856 -73.920760                               NaN   

         ... NUMBER_OF_PEDESTRIANS_KILLED NUMBER_OF_CYCLIST_INJURED  \
0        ...                            0                         0   
1        ...                            0                         0   
2        ...                            0                         0   
3        ...                            0                         0   
4        ...                            0                         0   
...      ...                          ...                       ...   
1918532  ...                            0                         0   
1918533  ...                            0                         0   
1918534  ...                            0                         0   
1918535  ...                            0                         0   
1918536  ...                            0                         0   

         NUMBER_OF_CYCLIST_KILLED  NUMBER_OF_MOTORIST_INJURED  \
0                               0                           0   
1                               0                           0   
2                               0                           1   
3                               0                           0   
4                               0                           0   
...                           ...                         ...   
1918532                         0                           0   
1918533                         0                           1   
1918534                         0                           1   
1918535                         0                           0   
1918536                         0                           1   

         NUMBER_OF_MOTORIST_KILLED   CONTRIBUTING_FACTOR_VEHICLE_1  \
0                                0                     Unspecified   
1                                0                Brakes Defective   
2                                0             Driver Inexperience   
3                                0                     Unspecified   
4                                0                     Unspecified   
...                            ...                             ...   
1918532                          0                     Unspecified   
1918533                  

preguntas:

    porcentaje de choques por día de la semana (por mes)                        #####################

    porcentaje de choques por hora (por día de la semana y mes)                 ################

    zona con mas choques                                                        ######################

    mes con mas choques?? (deberia agregar una columna con numero de mes)       ##################

    porcentaje de choques que resultan en muertes                               #####################
    
    porcentaje de choques por clima adverso

## porcentaje de choques por hora HISTORICO (8 años)

    como en este dataframe cada registro es un choque, y cada colision en teoria tiene un ID unico, puedo agrupar por horario del choque y hacer un conteo, entonces el numero bajo la columna "COLLISION_ID" es la cantidad de choques a esa hora en TODOS los registros dispoibles

In [20]:
df_time_perc_gen = df_vehicle_crash_clean.groupby(by = "CRASH_TIME").count()
df_time_perc_gen.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_time_perc_gen = df_time_perc_gen[["COLLISION_COUNT"]]

df_time_perc_gen

COLLISION_COUNT
CRASH_TIME                 
0:00                  18162
0:01                   2413
0:02                    494
0:03                    317
0:04                    310
...                     ...
9:55                   2500
9:56                    459
9:57                    473
9:58                    560
9:59                    305

[1440 rows x 1 columns]

In [21]:
df_time_perc_per_month = df_vehicle_crash_clean.groupby(by = ["CRASH_TIME", "MONTH"]).count()

df_time_perc_per_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_time_perc_per_month = df_time_perc_per_month[["COLLISION_COUNT"]]

df_time_perc_per_month

COLLISION_COUNT
CRASH_TIME MONTH                 
0:00       1                 1286
           2                 1142
           3                 1376
           4                 1292
           5                 1605
...                           ...
9:59       8                   28
           9                   22
           10                  24
           11                  22
           12                  21

[17278 rows x 1 columns]

In [22]:
df_time_perc_per_year = df_vehicle_crash_clean.groupby(by = ["CRASH_TIME", "YEAR"]).count()

df_time_perc_per_year.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_time_perc_per_year = df_time_perc_per_year[["COLLISION_COUNT"]]

df_time_perc_per_year

COLLISION_COUNT
CRASH_TIME YEAR                 
0:00       2012               12
           2013               25
           2014               20
           2015               24
           2016             2895
...                          ...
9:59       2018               38
           2019               42
           2020               19
           2021               27
           2022               17

[15836 rows x 1 columns]

In [33]:
df_time_perc_per_year_month = df_vehicle_crash_clean.groupby(by = ["CRASH_TIME", "YEAR", "MONTH"]).count()

df_time_perc_per_year_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_time_perc_per_year_month = df_time_perc_per_year_month[["COLLISION_COUNT"]]

df_time_perc_per_year_month

COLLISION_COUNT
CRASH_TIME YEAR MONTH                 
0:00       2012 7                    4
                8                    1
                9                    2
                10                   1
                11                   1
...                                ...
9:59       2022 3                    3
                4                    3
                5                    2
                6                    4
                8                    1

[157242 rows x 1 columns]

CRASH_TIME paso a ser el indice de este dataframe y muestro solo la columna de id de colisiones

In [23]:
# df_time_perc.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)
# df_time_perc = df_time_perc[["COLLISION_COUNT"]]

# df_time_perc

vemos que el horario con los mayores choques es a las 4PM, que coincide con la hora pico de trafico en la ciudad

las fechas tomadas son de 01/01/2013 hasta 12/31/2021

entre estas fechas hay 3286 dias, la cantidad de choques es de 27009, teniendo aproximadamente 8 choques todos los dias de los ultimos 8 años, a las 4 de la tarde

las 4PM tienen una presencia de 1.41% en el total de los horarios de choques, seguido de las 5PM con 1.38% y las 3PM con 1.37%

In [24]:
registros = len(df_vehicle_crash_clean["COLLISION_ID"]) #total de registros (1918537)

# registros                     ->      100%
# suma de colisiones por hora   ->      x

df_time_perc_gen["PERC"] = ( (df_time_perc_gen["COLLISION_COUNT"] * 100) / registros)

df_time_perc_gen = df_time_perc_gen[["COLLISION_COUNT", "PERC"]] #me quedo solo con las columnas que me interesan

df_time_perc_gen.sort_values("COLLISION_COUNT", ascending = False)

COLLISION_COUNT      PERC
CRASH_TIME                           
16:00                 27009  1.407791
17:00                 26471  1.379749
15:00                 26345  1.373182
18:00                 24453  1.274565
14:00                 24174  1.260023
...                     ...       ...
3:59                     84  0.004378
5:31                     84  0.004378
4:31                     80  0.004170
4:02                     80  0.004170
3:31                     80  0.004170

[1440 rows x 2 columns]

## COMVERTIR A CSV

In [135]:
# df_time_perc_gen.to_csv("kpi_time_crash_perc_gen.csv")
# df_time_perc_per_month.to_csv("kpi_time_crash_perc_per_month.csv")
# df_time_perc_per_year.to_csv("kpi_time_crash_perc_per_year.csv")
# df_time_perc_per_year_month.to_csv("kpi_time_crash_perc_per_year_month.csv")

los 3 horarios con menor cantidad de choques son:

    3:31AM
    4:02AM
    4:31AM

con una presencia de 0.0042% cada uno

In [25]:
df_time_perc_gen.sort_values("COLLISION_COUNT", ascending = False).tail(6)

COLLISION_COUNT      PERC
CRASH_TIME                           
3:41                     84  0.004378
3:59                     84  0.004378
5:31                     84  0.004378
4:31                     80  0.004170
4:02                     80  0.004170
3:31                     80  0.004170

#########################################################################################################################
## PUEDE SER DE INTERES

In [15]:
df_time_perc_test = df_vehicle_crash_clean.groupby(by = ["CRASH_TIME", "COLLISION_ID"]).sum()

In [16]:
df_time_perc_test #puede ser interesante

WEEK_DAY   LATITUDE  LONGITUDE  \
CRASH_TIME COLLISION_ID                                   
0:00       37776                4  40.761093 -73.983261   
           83539                6   0.000000   0.000000   
           196189               4  40.707623 -73.962131   
           212283               5  40.698151 -73.792404   
           307726               2  40.686206 -73.864397   
...                           ...        ...        ...   
9:59       4536981              6   0.000000   0.000000   
           4537559              2  40.746403 -73.979730   
           4539112              3   0.000000   0.000000   
           4541676              1  40.785150 -73.940630   
           4552998              6  40.815740 -73.954560   

                         NUMBER_OF_PERSONS_INJURED  NUMBER_OF_PERSONS_KILLED  \
CRASH_TIME COLLISION_ID                                                        
0:00       37776                               0.0                       0.0   
           83539                               0.0                       0.0   
           196189                              0.0                       0.0   
           212283                              0.0                       0.0   
           307726                              0.0                       0.0   
...                                            ...                       ...   
9:59       4536981                             0.0                       0.0   
           4537559                             0.0                       0.0   
           4539112                             0.0                       0.0   
           4541676                             1.0                       0.0   
           4552998                             0.0                       0.0   

                         NUMBER_OF_PEDESTRIANS_INJURED  \
CRASH_TIME COLLISION_ID                                  
0:00       37776                                     0   
           83539                                     0   
           196189                                    0   
           212283                                    0   
           307726                                    0   
...                                                ...   
9:59       4536981                                   0   
           4537559                                   0   
           4539112                                   0   
           4541676                                   0   
           4552998                                   0   

                         NUMBER_OF_PEDESTRIANS_KILLED  \
CRASH_TIME COLLISION_ID                                 
0:00       37776                                    0   
           83539                                    0   
           196189                                   0   
           212283                                   0   
           307726                                   0   
...                                               ...   
9:59       4536981                                  0   
           4537559                                  0   
           4539112                                  0   
           4541676                                  0   
           4552998                                  0   

                         NUMBER_OF_CYCLIST_INJURED  NUMBER_OF_CYCLIST_KILLED  \
CRASH_TIME COLLISION_ID                                                        
0:00       37776                                 0                         0   
           83539                                 0                         0   
           196189                                0                         0   
           212283                                0                         0   
           307726                                0                         0   
...                                            ...                       ...   
9:59       4536981                               0                         0   

################################################################################################################################

## porcentaje de choques por día de la semana HISTORICO (8 años)

    lunes       = 0
    martes      = 1
    miercoles   = 2
    jueves      = 3
    viernes     = 4
    sabado      = 5
    domingo     = 6

In [27]:
df_vehicle_crash_clean.head()

COLLISION_ID  CRASH_DATE CRASH_TIME  WEEK_DAY  MONTH  YEAR    BOROUGH  \
0        281035  01/01/2013       0:01         1      1  2013     QUEENS   
1         18025  01/01/2013       0:05         1      1  2013  MANHATTAN   
2         12712  01/01/2013       0:07         1      1  2013  MANHATTAN   
3         47038  01/01/2013       0:20         1      1  2013  MANHATTAN   
4        229479  01/01/2013       0:25         1      1  2013     QUEENS   

    LATITUDE  LONGITUDE                    ON_STREET_NAME  ...  \
0  40.761609 -73.882678  86 STREET                         ...   
1  40.757732 -73.996852  10 AVENUE                         ...   
2  40.720968 -73.982168  EAST HOUSTON STREET               ...   
3  40.772074 -73.949939  EAST 80 STREET                    ...   
4  40.683430 -73.826601  115 STREET                        ...   

  NUMBER_OF_PEDESTRIANS_KILLED NUMBER_OF_CYCLIST_INJURED  \
0                            0                         0   
1                            0                         0   
2                            0                         0   
3                            0                         0   
4                            0                         0   

   NUMBER_OF_CYCLIST_KILLED  NUMBER_OF_MOTORIST_INJURED  \
0                         0                           0   
1                         0                           0   
2                         0                           1   
3                         0                           0   
4                         0                           0   

   NUMBER_OF_MOTORIST_KILLED  CONTRIBUTING_FACTOR_VEHICLE_1  \
0                          0                    Unspecified   
1                          0               Brakes Defective   
2                          0            Driver Inexperience   
3                          0                    Unspecified   
4                          0                    Unspecified   

   CONTRIBUTING_FACTOR_VEHICLE_2  CONTRIBUTING_FACTOR_VEHICLE_3  \
0                    Unspecified                            NaN   
1                    Unspecified                    Unspecified   
2    Traffic Control Disregarded                            NaN   
3                    Unspecified                            NaN   
4                    Unspecified                    Unspecified   

   CONTRIBUTING_FACTOR_VEHICLE_4  CONTRIBUTING_FACTOR_VEHICLE_5  
0                            NaN                            NaN  
1                            NaN                            NaN  
2                            NaN                            NaN  
3                            NaN                            NaN  
4                    Unspecified                            NaN  

[5 rows x 25 columns]

In [28]:
df_day_perc_gen = df_vehicle_crash_clean.groupby(by = "WEEK_DAY").count()
df_day_perc_gen = df_day_perc_gen[["COLLISION_ID"]]

df_day_perc_gen.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_day_perc_gen

COLLISION_COUNT
WEEK_DAY                 
0                  274475
1                  283363
2                  281196
3                  287450
4                  306448
5                  257244
6                  228361

In [35]:
df_day_perc_gen["PERC"] = ( (df_day_perc_gen["COLLISION_COUNT"] * 100) / registros)

# la cantidad de registros es siempre la misma

# registros                     ->      100%
# suma de colisiones por dia    ->      x

df_day_perc_gen.sort_values("PERC")

COLLISION_COUNT       PERC
WEEK_DAY                            
6                  228361  11.902872
5                  257244  13.408342
0                  274475  14.306474
2                  281196  14.656793
1                  283363  14.769744
3                  287450  14.982771
4                  306448  15.973004

vemos que la presencia de cantidad de choques esta muy repartida entre todos los dias de la semana, siendo el dia con menor cantidad de choques el DOMINGO con 11.9% de presencia

el dia de la semana con mas choques de los ultimos 8 años es el dia VIERNES, con una presencia de 15.97%

In [29]:
df_day_perc_per_month = df_vehicle_crash_clean.groupby(by = ["WEEK_DAY", "MONTH"]).count()
df_day_perc_per_month = df_day_perc_per_month[["COLLISION_ID"]]

df_day_perc_per_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_day_perc_per_month

COLLISION_COUNT
WEEK_DAY MONTH                 
0        1                20177
         2                20304
         3                22381
         4                20986
         5                21398
...                         ...
6        8                20363
         9                20798
         10               19821
         11               18442
         12               19311

[84 rows x 1 columns]

In [30]:
df_day_perc_per_year = df_vehicle_crash_clean.groupby(by = ["WEEK_DAY", "YEAR"]).count()
df_day_perc_per_year = df_day_perc_per_year[["COLLISION_ID"]]

df_day_perc_per_year.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_day_perc_per_year

COLLISION_COUNT
WEEK_DAY YEAR                 
0        2012            14310
         2013            29453
         2014            29963
         2015            31059
         2016            32291
...                        ...
6        2018            26589
         2019            24702
         2020            13842
         2021            14288
         2022             8390

[77 rows x 1 columns]

In [34]:
df_day_perc_per_year_month = df_vehicle_crash_clean.groupby(by = ["WEEK_DAY", "YEAR", "MONTH"]).count()
df_day_perc_per_year_month = df_day_perc_per_year_month[["COLLISION_ID"]]

df_day_perc_per_year_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_day_perc_per_year_month

COLLISION_COUNT
WEEK_DAY YEAR MONTH                 
0        2012 7                 2660
              8                 2118
              9                 2024
              10                2445
              11                2174
...                              ...
6        2022 4                 1052
              5                 1443
              6                 1213
              7                 1312
              8                  225

[854 rows x 1 columns]

In [137]:
# df_day_perc.to_csv("kpi_day_crash_perc.csv")
# df_day_perc_per_month.to_csv("kpi_day_crash_perc_per_month.csv")
# df_day_perc_per_year.to_csv("kpi_day_crash_perc_per_year.csv")
# df_day_perc_per_year_month.to_csv("kpi_day_crash_perc_per_year_month.csv")

##########################################################################################################################

## zona con mas choques HISTORICO (8 años)

In [36]:
df_vehicle_crash_clean.columns

Index(['COLLISION_ID', 'CRASH_DATE', 'CRASH_TIME', 'WEEK_DAY', 'MONTH', 'YEAR',
       'BOROUGH', 'LATITUDE', 'LONGITUDE', 'ON_STREET_NAME',
       'CROSS_STREET_NAME', 'OFF_STREET_NAME', 'NUMBER_OF_PERSONS_INJURED',
       'NUMBER_OF_PERSONS_KILLED', 'NUMBER_OF_PEDESTRIANS_INJURED',
       'NUMBER_OF_PEDESTRIANS_KILLED', 'NUMBER_OF_CYCLIST_INJURED',
       'NUMBER_OF_CYCLIST_KILLED', 'NUMBER_OF_MOTORIST_INJURED',
       'NUMBER_OF_MOTORIST_KILLED', 'CONTRIBUTING_FACTOR_VEHICLE_1',
       'CONTRIBUTING_FACTOR_VEHICLE_2', 'CONTRIBUTING_FACTOR_VEHICLE_3',
       'CONTRIBUTING_FACTOR_VEHICLE_4', 'CONTRIBUTING_FACTOR_VEHICLE_5'],
      dtype='object')

In [37]:
df_vehicle_crash_clean

COLLISION_ID  CRASH_DATE CRASH_TIME  WEEK_DAY  MONTH  YEAR  \
0              281035  01/01/2013       0:01         1      1  2013   
1               18025  01/01/2013       0:05         1      1  2013   
2               12712  01/01/2013       0:07         1      1  2013   
3               47038  01/01/2013       0:20         1      1  2013   
4              229479  01/01/2013       0:25         1      1  2013   
...               ...         ...        ...       ...    ...   ...   
1918532       4492926  12/31/2021       9:00         4     12  2021   
1918533       4491580  12/31/2021       9:16         4     12  2021   
1918534       4491634  12/31/2021       9:17         4     12  2021   
1918535       4493202  12/31/2021       9:29         4     12  2021   
1918536       4491080  12/31/2021       9:50         4     12  2021   

           BOROUGH   LATITUDE  LONGITUDE                    ON_STREET_NAME  \
0           QUEENS  40.761609 -73.882678  86 STREET                          
1        MANHATTAN  40.757732 -73.996852  10 AVENUE                          
2        MANHATTAN  40.720968 -73.982168  EAST HOUSTON STREET                
3        MANHATTAN  40.772074 -73.949939  EAST 80 STREET                     
4           QUEENS  40.683430 -73.826601  115 STREET                         
...            ...        ...        ...                               ...   
1918532     QUEENS  40.769684 -73.821650                         32 AVENUE   
1918533        NaN  40.839970 -73.877920               EAST TREMONT AVENUE   
1918534     QUEENS        NaN        NaN                        138 Street   
1918535        NaN  40.882630 -73.839516                 BAYCHESTER AVENUE   
1918536      BRONX  40.816856 -73.920760                               NaN   

         ... NUMBER_OF_PEDESTRIANS_KILLED NUMBER_OF_CYCLIST_INJURED  \
0        ...                            0                         0   
1        ...                            0                         0   
2        ...                            0                         0   
3        ...                            0                         0   
4        ...                            0                         0   
...      ...                          ...                       ...   
1918532  ...                            0                         0   
1918533  ...                            0                         0   
1918534  ...                            0                         0   
1918535  ...                            0                         0   
1918536  ...                            0                         0   

         NUMBER_OF_CYCLIST_KILLED  NUMBER_OF_MOTORIST_INJURED  \
0                               0                           0   
1                               0                           0   
2                               0                           1   
3                               0                           0   
4                               0                           0   
...                           ...                         ...   
1918532                         0                           0   
1918533                         0                           1   
1918534                         0                           1   
1918535                         0                           0   
1918536                         0                           1   

         NUMBER_OF_MOTORIST_KILLED   CONTRIBUTING_FACTOR_VEHICLE_1  \
0                                0                     Unspecified   
1                                0                Brakes Defective   
2                                0             Driver Inexperience   
3                                0                     Unspecified   
4                                0                     Unspecified   
...                            ...                             ...   
1918532                          0                     Unspecified   
1918533                  

In [38]:
df_vehicle_crash_clean.BOROUGH.fillna("Unknown", inplace = True)

In [40]:
df_zone_perc_gen = df_vehicle_crash_clean.groupby("BOROUGH").count() 
#aca es conteo porque tengo que ver cuantas colisiones unicas hubo por barrio

df_zone_perc_gen.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

en esta tabla se puede ver que en los ultimos 8 años, la zona con menor cantidad de choques es staten island, con una presencia de 2.9% del total de los choques

mientras que brooklyn tiene una presencia de 21.79% del total de choques, siendo el barrio de Nueva York con mayor cantidad de choques en este periodo

In [42]:
df_zone_perc_gen = df_zone_perc_gen[["COLLISION_COUNT"]]

df_zone_perc_gen["PERC"] = ( (df_zone_perc_gen["COLLISION_COUNT"] * 100) / registros)

df_zone_perc_gen.sort_values("COLLISION_COUNT")

COLLISION_COUNT       PERC
BOROUGH                                  
STATEN ISLAND            55603   2.898198
BRONX                   194069  10.115468
MANHATTAN               301482  15.714161
QUEENS                  354517  18.478507
BROOKLYN                418027  21.788842
Unknown                 594839  31.004823

In [45]:
df_zone_perc_per_month = df_vehicle_crash_clean.groupby(by = ["BOROUGH", "MONTH"]).count()
df_zone_perc_per_month = df_zone_perc_per_month[["COLLISION_ID"]]

df_zone_perc_per_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_zone_perc_per_month

COLLISION_COUNT
BOROUGH MONTH                 
BRONX   1                15630
        2                14407
        3                16009
        4                14536
        5                16804
...                        ...
Unknown 8                53191
        9                51966
        10               54209
        11               50568
        12               50090

[72 rows x 1 columns]

In [46]:
df_zone_perc_per_year = df_vehicle_crash_clean.groupby(by = ["BOROUGH", "YEAR"]).count()
df_zone_perc_per_year = df_zone_perc_per_year[["COLLISION_ID"]]

df_zone_perc_per_year.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_zone_perc_per_year

COLLISION_COUNT
BOROUGH YEAR                 
BRONX   2012             9511
        2013            19768
        2014            19688
        2015            21258
        2016            21660
...                       ...
Unknown 2018            82330
        2019            74462
        2020            39228
        2021            38329
        2022            21264

[66 rows x 1 columns]

In [47]:
df_zone_perc_per_year_month = df_vehicle_crash_clean.groupby(by = ["BOROUGH", "YEAR", "MONTH"]).count()
df_zone_perc_per_year_month = df_zone_perc_per_year_month[["COLLISION_ID"]]

df_zone_perc_per_year_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_zone_perc_per_year_month

COLLISION_COUNT
BOROUGH YEAR MONTH                 
BRONX   2012 7                 1641
             8                 1665
             9                 1548
             10                1605
             11                1444
...                             ...
Unknown 2022 4                 2927
             5                 3171
             6                 3100
             7                 2936
             8                  982

[732 rows x 1 columns]

In [48]:
# df_zone_perc.to_csv("kpi_zone_crash_perc.csv")
# df_zone_perc_per_month.to_csv("kpi_zone_crash_perc_per_month.csv")
# df_zone_perc_per_year.to_csv("kpi_zone_crash_perc_per_year.csv")
# df_zone_perc_per_year_month.to_csv("kpi_zone_crash_perc_per_year_month.csv")

## la suma de los porcentajes da 69%, una posibilidad es dropear los registros que tengan NaN en la columna de BOROUGH para que el porcentaje llegue a 100%

##########################################################################################################################

## mes con mas choques?? (deberia agregar una columna con numero de mes)

In [59]:
df_month_perc = df_vehicle_crash_clean
df_month_perc.head()

COLLISION_ID  CRASH_DATE CRASH_TIME  WEEK_DAY    BOROUGH   LATITUDE  \
0        281035  01/01/2013       0:01         1     QUEENS  40.761609   
1         18025  01/01/2013       0:05         1  MANHATTAN  40.757732   
2         12712  01/01/2013       0:07         1  MANHATTAN  40.720968   
3         47038  01/01/2013       0:20         1  MANHATTAN  40.772074   
4        229479  01/01/2013       0:25         1     QUEENS  40.683430   

   LONGITUDE                    ON_STREET_NAME  \
0 -73.882678  86 STREET                          
1 -73.996852  10 AVENUE                          
2 -73.982168  EAST HOUSTON STREET                
3 -73.949939  EAST 80 STREET                     
4 -73.826601  115 STREET                         

                  CROSS_STREET_NAME OFF_STREET_NAME  ...  \
0  30 AVENUE                                    NaN  ...   
1  WEST 39 STREET                               NaN  ...   
2  RIDGE STREET                                 NaN  ...   
3  YORK AVENUE                                  NaN  ...   
4  107 AVENUE                                   NaN  ...   

   NUMBER_OF_CYCLIST_INJURED  NUMBER_OF_CYCLIST_KILLED  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   NUMBER_OF_MOTORIST_INJURED  NUMBER_OF_MOTORIST_KILLED  \
0                           0                          0   
1                           0                          0   
2                           1                          0   
3                           0                          0   
4                           0                          0   

   CONTRIBUTING_FACTOR_VEHICLE_1  CONTRIBUTING_FACTOR_VEHICLE_2  \
0                    Unspecified                    Unspecified   
1               Brakes Defective                    Unspecified   
2            Driver Inexperience    Traffic Control Disregarded   
3                    Unspecified                    Unspecified   
4                    Unspecified                    Unspecified   

   CONTRIBUTING_FACTOR_VEHICLE_3  CONTRIBUTING_FACTOR_VEHICLE_4  \
0                            NaN                            NaN   
1                    Unspecified                            NaN   
2                            NaN                            NaN   
3                            NaN                            NaN   
4                    Unspecified                    Unspecified   

  CONTRIBUTING_FACTOR_VEHICLE_5 MONTH  
0                           NaN     1  
1                           NaN     1  
2                           NaN     1  
3                           NaN     1  
4                           NaN     1  

[5 rows x 24 columns]

In [60]:
dates = df_month_perc["CRASH_DATE"]
idx = pd.PeriodIndex(dates, freq = "M") #configuro la frecuencia para que sea por MES
month = idx.month #asigno la configuracion de numero de MES
df_month_perc["MONTH"] = month

In [61]:
df_month_perc = df_month_perc[["COLLISION_ID", "MONTH"]]
df_month_perc.head()

COLLISION_ID  MONTH
0        281035      1
1         18025      1
2         12712      1
3         47038      1
4        229479      1

In [62]:
df_month_perc = df_month_perc.groupby("MONTH").count()


vemos que en los datos historicos de los ultimos 8 años, el mes con menor cantidad de choques fue FEBRERO, con una presencia del 7.18%

mientras que el mes con mayor cantidad de choques fue JULIO, con unapresencia del 9.19%

entiendo que este resultado se da dado que en nueva york, los meses de julio y agosto son meses de "vacaciones", por lo que hay mucho mas movimiento y, por los resultados, choques

In [65]:
df_month_perc.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_month_perc["PERC"] = ( (df_month_perc["COLLISION_COUNT"] * 100) / registros)

df_month_perc.sort_values("COLLISION_COUNT") #cantidad de choques por mes

COLLISION_COUNT      PERC
MONTH                           
2               137782  7.181618
4               142450  7.424928
1               149106  7.771859
3               153659  8.009176
11              162504  8.470204
12              163014  8.496787
5               163556  8.525038
9               165228  8.612187
6               165522  8.627511
8               168368  8.775854
10              170892  8.907412
7               176456  9.197425

In [ ]:
df_zone_perc_per_year_month = df_vehicle_crash_clean.groupby(by = ["BOROUGH", "YEAR", "MONTH"]).count()
df_zone_perc_per_year_month = df_zone_perc_per_year_month[["COLLISION_ID"]]

df_zone_perc_per_year_month.rename(columns = {"COLLISION_ID" : "COLLISION_COUNT"}, inplace = True)

df_zone_perc_per_year_month

In [139]:
# df_month_perc.to_csv("kpi_month_crash_perc.csv")

##########################################################################################################################

## porcentaje de choques que resultan en muertes

In [66]:
df_vehicle_crash_clean.columns

Index(['COLLISION_ID', 'CRASH_DATE', 'CRASH_TIME', 'WEEK_DAY', 'BOROUGH',
       'LATITUDE', 'LONGITUDE', 'ON_STREET_NAME', 'CROSS_STREET_NAME',
       'OFF_STREET_NAME', 'NUMBER_OF_PERSONS_INJURED',
       'NUMBER_OF_PERSONS_KILLED', 'NUMBER_OF_PEDESTRIANS_INJURED',
       'NUMBER_OF_PEDESTRIANS_KILLED', 'NUMBER_OF_CYCLIST_INJURED',
       'NUMBER_OF_CYCLIST_KILLED', 'NUMBER_OF_MOTORIST_INJURED',
       'NUMBER_OF_MOTORIST_KILLED', 'CONTRIBUTING_FACTOR_VEHICLE_1',
       'CONTRIBUTING_FACTOR_VEHICLE_2', 'CONTRIBUTING_FACTOR_VEHICLE_3',
       'CONTRIBUTING_FACTOR_VEHICLE_4', 'CONTRIBUTING_FACTOR_VEHICLE_5',
       'MONTH'],
      dtype='object')

NUMBER_OF_PERSONS_KILLED, NUMBER_OF_PEDESTRIANS_KILLED, NUMBER_OF_CYCLIST_KILLED, NUMBER_OF_MOTORIST_KILLED

In [69]:
df_death_perc = df_vehicle_crash_clean[["COLLISION_ID", "NUMBER_OF_PERSONS_KILLED", "NUMBER_OF_PEDESTRIANS_KILLED", "NUMBER_OF_CYCLIST_KILLED", "NUMBER_OF_MOTORIST_KILLED"]]

In [77]:
df_death_perc = df_death_perc[df_death_perc["NUMBER_OF_PERSONS_KILLED"] > 0]

df_death_perc #en este dataframe tengo todos los registros en los que murio ALGUIEN

dicc_death = dict()

dicc_death["TOT_CRASHES"] = registros
dicc_death["TOT_DEATHS"] = len(df_death_perc["NUMBER_OF_PERSONS_KILLED"])

dicc_death

{'TOT_CRASHES': 1918537, 'TOT_DEATHS': 2537}

vemos que el porcentaje del total de los choques que resultaron en al menos una muerte es de un 0.13%, siendo 2537 choques que resultaron en muerte de al menos un participante

In [78]:
dicc_death["PERC_DEATH"] = ( (dicc_death["TOT_DEATHS"] * 100) / dicc_death["TOT_CRASHES"])

dicc_death

{'TOT_CRASHES': 1918537, 'TOT_DEATHS': 2537, 'PERC_DEATH': 0.13223617787929032}

In [129]:
df_death_perc = pd.DataFrame.from_dict(dicc_death, orient = "index").rename(columns = {0 : "VALUE"})
df_death_perc.round(3)

VALUE
TOT_CRASHES  1918537.000
TOT_DEATHS      2537.000
PERC_DEATH         0.132

In [140]:
# df_death_perc.to_csv("kpi_death_crash_perc.csv")

#############################################################################################################################################################

## porcentaje de choques por clima adverso

In [79]:
df_vehicle_crash_clean.CONTRIBUTING_FACTOR_VEHICLE_1.unique()

array(['Unspecified', 'Brakes Defective', 'Driver Inexperience',
       'Fatigued/Drowsy', 'Outside Car Distraction', 'Turning Improperly',
       'Traffic Control Disregarded', 'Driver Inattention/Distraction',
       'Failure to Yield Right-of-Way', 'Aggressive Driving/Road Rage',
       'Physical Disability', 'Glare', 'Oversized Vehicle',
       'Reaction to Uninvolved Vehicle', 'Other Vehicular',
       'Prescription Medication', 'Lost Consciousness',
       'Backing Unsafely', 'View Obstructed/Limited', 'Unsafe Speed',
       'Alcohol Involvement', 'Tire Failure/Inadequate',
       'Failure to Keep Right', 'Passenger Distraction',
       'Other Electronic Device', 'Steering Failure', 'Illness',
       'Pavement Slippery', 'Traffic Control Device Improper/Non-Working',
       'Pavement Defective', 'Unsafe Lane Changing', nan,
       'Drugs (illegal)', 'Animals Action', 'Obstruction/Debris',
       'Cell Phone (hand-held)', 'Fell Asleep', 'Windshield Inadequate',
       'Passing or 

se puede interpretar que "Pavement Slippery" puede ser una cuestion climatica, sea por lluvia o nieve, analizo...

In [85]:
df_climate_perc = df_vehicle_crash_clean[ (df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_1"] == "Pavement Slippery") | (df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_2"] == "Pavement Slippery") | (df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_3"] == "Pavement Slippery") | (df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_4"] == "Pavement Slippery") | (df_vehicle_crash_clean["CONTRIBUTING_FACTOR_VEHICLE_5"] == "Pavement Slippery")]

In [86]:
df_climate_perc = df_climate_perc[["COLLISION_ID", "CONTRIBUTING_FACTOR_VEHICLE_1", "CONTRIBUTING_FACTOR_VEHICLE_2", "CONTRIBUTING_FACTOR_VEHICLE_3", "CONTRIBUTING_FACTOR_VEHICLE_4", "CONTRIBUTING_FACTOR_VEHICLE_5"]]

df_climate_perc #al menos uno de los factores de contribucion fue por pavimento resbaladizo

COLLISION_ID CONTRIBUTING_FACTOR_VEHICLE_1  \
522             16269             Pavement Slippery   
633           3034605       Outside Car Distraction   
1133          3163602             Pavement Slippery   
1134          3145893             Pavement Slippery   
1344          3367387           Driver Inexperience   
...               ...                           ...   
1918275       4380680             Pavement Slippery   
1918303       4380451             Pavement Slippery   
1918490       4490809             Pavement Slippery   
1918494       4491652             Pavement Slippery   
1918508       4490825             Pavement Slippery   

        CONTRIBUTING_FACTOR_VEHICLE_2 CONTRIBUTING_FACTOR_VEHICLE_3  \
522                       Unspecified                           NaN   
633                 Pavement Slippery                           NaN   
1133                      Unspecified                           NaN   
1134                Pavement Slippery                   Unspecified   
1344                Pavement Slippery                           NaN   
...                               ...                           ...   
1918275             Pavement Slippery          Unsafe Lane Changing   
1918303                   Unspecified                   Unspecified   
1918490                           NaN                           NaN   
1918494                   Unspecified                           NaN   
1918508             Pavement Slippery                   Unspecified   

        CONTRIBUTING_FACTOR_VEHICLE_4 CONTRIBUTING_FACTOR_VEHICLE_5  
522                               NaN                           NaN  
633                               NaN                           NaN  
1133                              NaN                           NaN  
1134                      Unspecified                           NaN  
1344                              NaN                           NaN  
...                               ...                           ...  
1918275                           NaN                           NaN  
1918303                           NaN                           NaN  
1918490                           NaN                           NaN  
1918494                           NaN                           NaN  
1918508                   Unspecified                           NaN  

[20225 rows x 6 columns]

vemos que la presencia de factor por clima es del 1.054% entre todos los choques historicos (ultimos 8 años)

In [87]:
dicc_climate = dict()

dicc_climate["TOT_CRASHES"] = registros
dicc_climate["TOT_CLIMATE"] = len(df_climate_perc["COLLISION_ID"])

dicc_climate["PERC_CLIMATE"] = ( (dicc_climate["TOT_CLIMATE"] * 100) / dicc_climate["TOT_CRASHES"])

dicc_climate

{'TOT_CRASHES': 1918537,
 'TOT_CLIMATE': 20225,
 'PERC_CLIMATE': 1.0541886864835028}

In [131]:
df_climate_perc = pd.DataFrame.from_dict(dicc_climate, orient = "index").rename(columns = {0 : "VALUE"})

In [133]:
df_climate_perc.round(3)

VALUE
TOT_CRASHES   1918537.000
TOT_CLIMATE     20225.000
PERC_CLIMATE        1.054

In [141]:
# df_climate_perc.to_csv("kpi_climate_crash_perc.csv")